### ISSUE TITLE
Convert Hare to use Haiku's Layout Manager

### ISSUE THREAD

**Issue by soakbot** [HaikuArchives/Hare#1]
**Creator: Scott McCreary**
**URL: http://dev.osdrawer.net/issues/289**

The main window seems to keep getting bigger every time you open it up, this needs to be fixed.
Also there's buttons that are sometimes overlapping that also need to be fixed.

_CAUTION: comments not extracted from issues. See OsDrawer or the WaybackMachine for issue comments._

**Comment by scottmc** [comment]
This should really just be converted to use Haiku's Layout Manager instead.

**Comment by scottmc** [comment]
This may also solve issue #6.

**Comment by CodeforEvolution** [comment]
Hello! This is CodeforEvolution from GCI! Sorry I could not submit something on time as I got into a very busy situation, otherwise, when I do have time, I will continue working on this.

**Comment by CodeforEvolution** [comment]
Hey, as a status update I have been stalled with a few issues for quite some time:

- Dragging and dropping items into the actual column list view does not work
- When dragging items into another part of the app window an entry is created in the list, though no info is added
- The app's functionality overall seems very broken
- I still have to convert the app to layouting, only have removed libsanta stuff so far and still fixing from there

To see my progress so far:
https://github.com/CodeforEvolution/Hare/tree/layouting

**Comment by smccrearyx** [comment]
Nice progress.  Feel free to submit code as pull requests as you progress thru the various open issues.

**Comment by CodeforEvolution** [comment]
Thanks! The only problem is the application is too unstable and unstable at the moment, and it basically has no functionality, I was wondering if you’d have any idea on fixing the above problems in the previous comment I made, as I have been stuck on those problems for a long time.

**Comment by scottmc** [comment]
Are you in GCI this year?  If so there is an open task to fix an issue in Hare.  Perhaps send your pending changes as a new pull request, and then after it is merged open new issues for each of the pending issues you noted up above.  We can create new instances for students who want to fix more than one issue per HaikuArchive app, just ask us to open another one.

**Comment by CodeforEvolution** [comment]
Yes, I have unfortunately been quite busy though, and if anything should have time to merge in the upcoming weekend.

**Comment by scottmc** [comment]
ok, cool.  I added a task for issue #2, as that's probably the biggest pending issue after fixing the layout.

**Comment by scottmc** [comment]
Turns out another student chose to work on the layout issue and has now submitted a pull request.  This will likely conflict with @CodeforEvolution 's unsubmitted code.

**Comment by scottmc** [comment]
![markup1](https://user-images.githubusercontent.com/1039279/34196379-bdcd482e-e516-11e7-8f3e-2732684c6009.png)
Here's a markup of my vision of what this should be like.  Move genre and comment on r and under year.  Have three views with adjustable split bars between then, defaulted to center like shown.  Fix window resizing, I recall disabling it at some point since it was causing the window to grow out of control.  Make all three views have option for vertical scroll view in case contents fill up box.  Fix upper left box to also use layout manager. Fix bottom status area so that it all fits in there nicer, looks cramped right now. Make text entry boxes wider, perhaps let them grow with the width of each view.

**Comment by scottmc** [comment]
@CodeforEvolution we will likely merge @friedrich12 's Pull Request once it gets updated to fix listed issues.  This will likely mess up your pending work.  I do like your work to Completely Redo Build System and Directory Structure, perhaps you can redo that once the layout manager coding in done.  With the pending pull request it can already allow you to drag a CD onto the bottom view and populate it, so looking promising.

**Comment by CodeforEvolution** [comment]
Of course, I will watch for the completed pull request, and then rebase my code once the time comes.

**Comment by scottmc** [comment]
@CodeforEvolution code has now been merged.  If you want to just apply the changes you made to remove libsanta and taglib from the repo, feel free to do so.  I believe there is an open task to pick any open issue for Hare and solve it.  That way at least part of your work will still be useful

@All This issue is about half way done.  The red marks on the markup image above still need to be fixed up.  If you take on the task, don't be afraid to ask questions.

**Comment by friedrich12** [comment]
I plan on finishing most of the UI today if not all the remaining fixes. Also can you give me some information about adding scrollbars to the window? Please also show me any coding style that must be followed if needed?

**Comment by humdingerb** [comment]
https://www.haiku-os.org/development/ has links to information, like the Haiku Coding Guidelines and esp. the BeBook and HaikuBook that describe the API. Scrollbars are in the interface kit.
See also http://xref.plausible.coop/source/ to search through Haiku's source for examples (under apps and preferences).

**Comment by CodeforEvolution** [comment]
Hey,if all the layout changes are done, I can start reworking my code base.

**Comment by scottmc** [comment]
i think @friedrich12 was going to work on finishing up the UI changes to get it to look and work like markup.  @CodeforEvolution you claimed the part 2 task, if yu intend to finished up the UI that is fine.  Or you can drop that task and pick the other Hare task to "fix an open issue" on Hare and then work on the code base clean up of removing lib santa, tablib, etc. But you might want to wait to start that after the UI code is completed which might go quick if @friedrich was working on it on the 22nd.  Tricky to work on a smaller project in parallel, so if you could wait a couples days that might work out.

**Comment by CodeforEvolution** [comment]
Sorry about that, I will drop that task, just @ me when I can jump in.

**Comment by scottmc** [comment]
@friedrich12 any progress on this?  If so you may want to put a claim in on the task.

**Comment by scottmc** [comment]
@CodeforEvolution feel free to jump in I guess.  Be sure to use the current code.

**Comment by CodeforEvolution** [comment]
Hello! I have updated the interface quite a bit and looks now quite nice! I did find though that BScrollViews weren't actually useful with the amount of content...otherwise, should I create a pull request? You can see where I'm at here: https://www.github.com/CodeforEvolution/Hare, (use the final-layouting branch).

**Comment by scottmc** [comment]
Keep in mind that the amount of content is likely to grow once we expand to allow for more tags.  Yes, once you are happy with the code do a pull request

**Comment by CodeforEvolution** [comment]
True, for now though, I found it to be very buggy with drawing...leaving lots of artifacts of the scroll views' contents.

**Comment by CodeforEvolution** [comment]
Well, I'll create a pull request!

**Comment by scottmc** [comment]
You are now almost there, just need to get the bottom section working again.

**Comment by scottmc** [comment]
New layout almost there.  Is it possible to move the Apply buttoms to be in bottom right, but under the last row in the same column and not the next column over?  This would allow the text entry boxes for both sections to use the full width.  The drag and drop and loading of the CD seems to still be broken.  But at least now the bottom view is back, just not functional yet.  Overall the window is now working mostly as I envisioned it.  Is it possible to make the divider bars moveable as well?  I'm not converned with the ghosting issues, as most users will adjust the window to their liking and leave it that way.  The default open location needs to be adjusted, right now it seems to be set to 0,0 which puts the tab off the screen.  Perhaps check others apps to see where most of them set the default open position as.  Nice progress.

**Comment by janus2** [comment]
The ghosting problem is related to https://dev.haiku-os.org/ticket/3037.

**Comment by CodeforEvolution** [comment]
I'm now so close to finishing (Sorry for how long it's been taking, I have been busy). The only problem now is that the info from added media files aren't being added to the BColumnListView. I've used Debugger and traced through many variables, but I can't figure out why the information isn't appearing of the column list view.

**Comment by janus2** [comment]
When I say BColumnListView is a bitch nobody listen to me ;-)
I suggest you try to add this lines in your BRefRow constructor and see the evil in the eyes:

`SetField(new BBitmapField(NULL), 0);`
`for (int i=0; i < 10; i++)`
`   SetField(new BStringField(B_EMPTY_STRING), i);`

**Comment by CodeforEvolution** [comment]
No luck unfortunatly...any other ideas?

**Comment by janus2** [comment]
I tested the code it works...

**Comment by janus2** [comment]
I don't have cd driver I used a folder full of songs can be this the  difference?

**Comment by CodeforEvolution** [comment]
Oh...let me double check something then...

**Comment by CodeforEvolution** [comment]
Well, I tried that too, what version of the Hare code are you using?

**Comment by janus2** [comment]
Master of this repo

![screenshot](https://user-images.githubusercontent.com/10423645/36758840-d570621e-1c15-11e8-80f8-0dcca8387fb2.png)

**Comment by CodeforEvolution** [comment]
Something must be wrong with my computer then...I tried the master just now and had no dice...what hrev of Haiku are you using? I'm using hrev 51818.

**Comment by janus2** [comment]
hrev 51792, but the BColumnListView is unchanged since than.

**Comment by CodeforEvolution** [comment]
Let me try another computer...

**Comment by janus2** [comment]
Don't you have a release and a debug version? Maybe you are using an old version of the app, often this happen to me.

**Comment by janus2** [comment]
I have update my system to hrev 51818 and it works.

**Comment by janus2** [comment]
I have add a PR you can verify with that version.

**Comment by CodeforEvolution** [comment]
It does work! I must've fixed the issue, and then some how broken it again...

**Comment by janus2** [comment]
Good!

**Comment by scottmc** [comment]
i suspect a pull request will happen soon then?

**Comment by CodeforEvolution** [comment]
Yes, a pull request will come in soon. (I have been a bit busy lately)

**Comment by scottmc** [comment]
Fixed in 05500f33d667fc530562bfca6cea8392d1afd3a4
